### PR TITLE
Add possibility to exclude single joysticks or midi devices

### LINKED ISSUE #1554: Add possibility to exclude single joysticks or midi devices
**Is your feature request related to a problem? Please describe.**
If a midi device is controlled by another application, it must be excluded from mobiflight. 
Also unneeded joysticks can be excluded.

**Describe the solution you'd like**
Add possibility to exclude single joysticks or midi devices in peripherals settings dialog.

**Additional context**
Is related to #1419.

### PR DESCRIPTION
**Pull request by Koseng**
fixes #1554 
related to #1419 
- [x] i18n
- [x] ~~unit tests~~
- [ ] documentation

image

### CODE REVIEW

File: MobiFlight/Joysticks/JoystickManager.cs [hunk 1]
```diff
@@ -121,7 +128,9 @@ public void SetHandle(IntPtr handle)
         public void Connect()
         {
             var di = new SharpDX.DirectInput.DirectInput();
-            joysticks?.Clear();
+            Joysticks?.Clear();
+            ExcludedJoysticks?.Clear();
+            string[] excludedJoysticks = Properties.Settings.Default.ExcludedJoysticks.Split('§');
```
Comment (DocMoebiuz) [inline MobiFlight/Joysticks/JoystickManager.cs:133]: we never used § as a delimeter yet. You can actually serialize the data structure like I did in MainForm for the autoload-configs... i think that is way nicer.

see MainForm > RestoreAutoLoadConfig & SaveAutoLoadConfig

File: UI/Panels/Settings/PeripheralsPanel.cs [hunk 2]
```diff
@@ -16,50 +18,81 @@ public PeripheralsPanel()
...
+            bool enableJoysticksChanged = Properties.Settings.Default.EnableJoystickSupport != checkBoxJoystickSupport.Checked;
+            bool enableMidiChanged = Properties.Settings.Default.EnableMidiSupport != checkBoxMidiSupport.Checked;
+
+            Properties.Settings.Default.ExcludedMidiBoards = string.Join("§", excludedBoards);
+            Properties.Settings.Default.ExcludedJoysticks = string.Join("§", excludedJoysticks);
+            Properties.Settings.Default.EnableJoystickSupport = checkBoxJoystickSupport.Checked;
+            Properties.Settings.Default.EnableMidiSupport = checkBoxMidiSupport.Checked;
+
+            // Update joystick management
+            if (!checkBoxJoystickSupport.Checked && enableJoysticksChanged)
```
Comment (DocMoebiuz) [inline UI/Panels/Settings/PeripheralsPanel.cs:77]: Ideally these actions should not happen withing the settingsdialog
We have a settings changed event handler that could handle this. We can do that at a later point... but it makes sense as refactoring.

**Review by DocMoebiuz — CHANGES_REQUESTED** [review]

File: MobiFlight/Joysticks/JoystickManager.cs (hunk 1 above)
Comment (Koseng) [reply MobiFlight/Joysticks/JoystickManager.cs]: yeah that's nicer. Done.

File: UI/Panels/Settings/PeripheralsPanel.cs (hunk 2 above)
Comment (Koseng) [reply UI/Panels/Settings/PeripheralsPanel.cs:78]: Yes, with settings changed event is better. I have put it on my todo list for later.

File: MobiFlight/Joysticks/JoystickManager.cs [hunk 3]
```diff
@@ -97,20 +98,26 @@ public void Startup()
         public void Shutdown()
         {
             PollTimer.Stop();
-            joysticks.Clear();
+            Joysticks.Clear();
```
Comment (DocMoebiuz) [inline MobiFlight/Joysticks/JoystickManager.cs:101]: I think Shutdown should call Stop for all Joysticks to make sure, that they can do their specific stuff. Like turning off the lights or disconnecting from their stream (Octavi)

Comment (DocMoebiuz) [reply MobiFlight/Joysticks/JoystickManager.cs:101]: But tbh, we have to provide a Shutdown or Disconnect method for HID devices. In the case of Octavi it opens it's own stream. That's why it keeps sending events even after Shutdown of the JoystickManager

**Review by DocMoebiuz — CHANGES_REQUESTED** [review]
See my comments.

**Review by DocMoebiuz — APPROVED** [review]
Thanks!

**Pull request merged**